zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkConnection.java

```diff
@@ -192,4 +230,53 @@ public String getServers() {
   public void addAuthInfo(String scheme, byte[] auth) {
     _zk.addAuthInfo(scheme, auth);
   }
+
+  private void lookupGetChildrenMethod() {
+    if (_getChildrenMethod != null) {
+      // Method is already cached.
+      return;
+    }
+    try {
+      if (GETCHILDREN_PAGINATION_DISABLED) {
+        lookupNonPaginatedGetChildren();
+      } else {
+        // Lookup the paginated getChildren API
+        _getChildrenMethod =
+            ZooKeeper.class.getMethod("getAllChildrenPaginated", String.class, boolean.class);
+      }
+    } catch (NoSuchMethodException e1) {
+      // Pagination API is not supported, fall back to non-paginated API
+      lookupNonPaginatedGetChildren();
+    }
+    LOG.info("Pagination config {}={}, method to be invoked: {}",
+        ZkSystemPropertyKeys.ZK_GETCHILDREN_PAGINATION_DISABLED, GETCHILDREN_PAGINATION_DISABLED,
+        _getChildrenMethod.getName());
+  }
+
+  private void lookupNonPaginatedGetChildren() {
+    try {
+      _getChildrenMethod = ZooKeeper.class.getMethod("getChildren", String.class, boolean.class);
+    } catch (NoSuchMethodException e2) {
+      // We should not expect this exception here.
+      LOG.error("getChildren is not supported in this zookeeper version!");
+    }
+  }
+
+  private void handleInvokedMethodException(Throwable cause)
```

Comment:
Just to confirm, as we discussed, if the ZK server is upgraded in between, the client shall be able to leverage the new API immediately, right? In this case, I assume we expect the ZkConnection object to be recreated. And I believe this is the case when we handle new session creation.

But what if the server upgrade is fast and the session is not expired? In this case, since you update the method here (and fallback to the older one), then the client will never use pagination call unless we restart the client.

Reply:
> Just to confirm, as we discussed, if the ZK server is upgraded in between, the client shall be able to leverage the new API immediately, right? In this case, I assume we expect the ZkConnection object to be recreated. And I believe this is the case when we handle new session creation.

If zk server is upgraded after the client has been created, we can not guarantee the client will always know server has enabled pagination. The example is what you said "what if the server upgrade is fast and the session is not expired". If session expires during the upgrade, client can reset the cache and re-detect the pagination. It cannot solving the problem of completely auto detecting.

> But what if the server upgrade is fast and the session is not expired? In this case, since you update the method here (and fallback to the older one), then the client will never use pagination call unless we restart the client.

ZkClient cannot automatically and dynamically handle this case. The agreement is, to enable pagination, ZK server should be supposed to upgrade to have pagination as the first step. If client has pagination first but server upgrades later, client should be expected to bounce.
To detect the server has enabled the pagination without a session change, every time the client's getChildren call must try pagination first. This is inefficient and not what we want. We have to trade off. And obviously we'd expect clients to restart if this case happens.

Reply:
One possible way is that, once the call throws an unsupported exception, we fallback to the old getChildren method.

Then on every new session establishment, we reset it based on the configuration and zk client lib method availability.

Reply:
This can only solve the case when session is re-established, not the case when server is upgrade but no session re-establishing. Yes, we've invalidated the cache each time session is changed.